Sources/MapboxDirections/RouteRefreshResponse.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+
+public struct RouteRefreshResponse {
+    public let httpResponse: HTTPURLResponse?
+    
+    public let identifier: String?
+    public var routeIndex: Int
+    public var legIndex: Int
+    
+    public var route: Route? // pass it in a user info? and then consturct new route during decoding?
+    
+    public let credentials: DirectionsCredentials
+    
+    private var legAnnotations: [RouteLegAnnotation]
+    
+    /**
+     The time when this `RouteRefreshResponse` object was created, which is immediately upon recieving the raw URL response.
+     
+     If you manually start fetching a task returned by `Directions.url(forCalculating:)`, this property is set to `nil`; use the `URLSessionTaskTransactionMetrics.responseEndDate` property instead. This property may also be set to `nil` if you create this result from a JSON object or encoded object.
+     
+     This property does not persist after encoding and decoding.
+     */
+    public var created: Date = Date()
+}
+
+extension RouteRefreshResponse: Codable {
+    enum CodingKeys: String, CodingKey {
+        case identifier = "uuid"
+        case route
+    }
+    
+    private func applyRefreshingTo(route: Route) -> Route{
+        let updatedLegs = zip(legAnnotations, route.legs.enumerated()).map { (pair) -> RouteLeg in
+            let (annotation, legEnum) = pair
+            let leg = legEnum.element.copy() as! RouteLeg
+            if legEnum.offset >= legIndex {
+                leg.updateAnnotationData(from: annotation)
+            }
+            return leg
+        }
+        
+        let route = Route(legs: updatedLegs,
+                           shape: route.shape,
+                           distance: route.distance,
+                           expectedTravelTime: route.expectedTravelTime)
```

Comment:
Instead of creating a brand-new Route object based on the old one, can we mutate the existing route object? I’m just concerned that some SDK or application code may have stored a reference to the old Route that’s now outdated. It would be fine to make additional properties read-write in order to reuse the existing data structures. If we’re concerned about changing data out from behind the client’s back, then I suppose we could create a parallel MutableRoute class instead, but that could be a lot to maintain. In the future, once we find a way to convert these types to structs, we’ll be able to make mutability more explicit anyways.

Reply:
Making `Route` mutable should still be done explicitly outside `Directions.refresh` because (1) users should be aware of the update and have control over it , (2) refreshing route also involves updating `RouteProgress` (and potentially anything else) and doing some calculations, so it is safer be done by user's completion.

Reply:
> users should be aware of the update and have control over it

Developers can observe changes to the Route and RouteLeg properties using key-value observation. The refreshable properties are already individually mutable, so there’s already an expectation that they could change behind the developer’s back. What the interface doesn’t explicitly call out is that refreshing the route silently makes any existing route references become stale; client code might not be in a good position to obtain the refreshed route in that case.